OpenShift 201 Resource Management (#116)
BCDevOps/devops-platform-workshops: openshift-201/resource-mgmt.md

* If the memory allocated by all of the processes in a container exceeds the memory limit, the node Out of Memory (OOM) killer will immediately select and kill a process in the container.

Resource request and resource limits should be defined for each container in either a Deployment, DeploymentConfiguration, StatefulSets, BuildConfigs, and CronJob. If requests and limits have not been defined, then you will find a resources: {} line for each container.

> Lets create a deployment to test! Create this deployment in your project.

Reviewer: ```suggestion
Let's create a deployment to test! Create this deployment in your project.
```

          - name: SERVICE_PORT
            value: "443"
          - name: REQUESTS
            value: "500000"
          - name: CONCURRENCY
            value: "20"
          command: ["/opt/rh/httpd24/root/usr/bin/ab"]
          args: ["-dSrk", "-c $(CONCURRENCY)", "-n $(REQUESTS)", "https://$(SERVICE_HOST):$(SERVICE_PORT)/index.html"]
  
```

From the web console if you change to developer view and navigate to the Monitoring tab select your nginx deployment. You should see the load-test pod traffic increasing metrics for our pod.

![cpu load](images/resource-mgmt/pod-load-cpu.png) 

> From the web console select your hello-world-nginx pod and navigate to the Metrics tab. We can see the traffic we are sending our pod is affecting the cpu quite a bit. In this example we can see the actual cpu usage is well over the request we set and  over 100% of the limit we set.

Reviewer 2: Too many spaces
```suggestion
From the web console select your hello-world-nginx pod and navigate to the Metrics tab. We can see the traffic we are sending our pod is affecting the cpu quite a bit. In this example we can see the actual cpu usage is well over the request we set and over 100% of the limit we set.
```

![cpu quota](images/resource-mgmt/pod-load-cpu-quota.png)

Because the actual cpu usage is higher than our cpu limit openshift/kubernetes will throttle the available cpu to our pod. This would affecting the performance of our web server and cause response times of our application.

![cpu throttle](images/resource-mgmt/pod-load-cpu-throttle.png)

You can create a custom PromQL query to view the CPU throttling by using this query and updating your pod name and namespace name:

> ```

Reviewer 2: Add syntax highlighting
```suggestion
```GraphQL
```

4. Set container cpu/memory limit.

> Setting a limit has the effect of immediately killing a container process or cpu throttling if the combined cpu or memory usage of all processes in the container exceeds the limit, and is therefore a mixed blessing. On the one hand, it may make unanticipated excess cpu/memory usage obvious early ("fail fast"); on the other hand it also terminates processes abruptly.

Author: Hmmm. Ya I see your point. More I wanted to call out that the limit can also catch any bad code or mis-configuration that would cause memory or cpu to spike in the container so you catch it early and not like weeks down the road do you find a memory leak bug as the limit was too high. Also calling out that setting a limit will start killing processes or throttling things if set too low.

Happy for a suggestion here though!

> Setting a limit has the effect of immediately killing a container process or cpu throttling if the combined cpu or memory usage of all processes in the container exceeds the limit, and is therefore a mixed blessing. On the one hand, it may make unanticipated excess cpu/memory usage obvious early ("fail fast"); on the other hand it also terminates processes abruptly.

Reviewer 3: Wondering if there is a way of describing this with a 'goldilocks' example, might help to emphasize a happy middle ground?:
- If the limit is set too low =... 
- If the limit is set too high/not set =... 
- If the limit is set _just right_ =...

> Setting a limit has the effect of immediately killing a container process or cpu throttling if the combined cpu or memory usage of all processes in the container exceeds the limit, and is therefore a mixed blessing. On the one hand, it may make unanticipated excess cpu/memory usage obvious early ("fail fast"); on the other hand it also terminates processes abruptly.

Author: Great idea! I added a summary to this section. Let me know if that works.

> Setting a limit has the effect of immediately killing a container process or cpu throttling if the combined cpu or memory usage of all processes in the container exceeds the limit, and is therefore a mixed blessing. On the one hand, it may make unanticipated excess cpu/memory usage obvious early ("fail fast"); on the other hand it also terminates processes abruptly.

Reviewer 3: Awesome, thanks Phil. Looks good to me!

| Scope | Description|
|---|---|
|Terminating | Match pods where spec.activeDeadlineSeconds >= 0.|
| NotTerminating | Match pods where spec.activeDeadlineSeconds is nil.|
| BestEffort | Match pods that have best effort quality of service for either cpu or memory.|
| NotBestEffort| Match pods that do not have best effort quality of service for cpu and memory.|

You'll notice there are a few quotas within the BC Gov OpenShift projects. Have a look at quota details and see how much has been consumed.
## Limit Ranges

A LimitRange resource, also called a limit, defines the default, minimum, and maximum values for compute resource requests, and the limits for a single pod or container defined inside the project. 

To understand the difference between a limit range and a resource quota, consider that a limit range defines valid ranges and default values for a single container, and a resource quota defines only top values for the sum of all pods in a project. 

A cluster administrator concerned about resource usage in an OpenShift cluster usually defines both limits and quotas for a project.

> A limit range resource can also define minimum, and maximum values for the storage capacity requested by an image, image stream, or persistent volume claim. 

Reviewer 2: ```suggestion
A limit range resource can also define minimum, and maximum values for the storage capacity requested by a persistent volume claim; the maximum size of an image; or the number of image streams and tags.
```

The following table describes some resources that a quota can restrict by their count or number:

|Resource Name|	Quota Description|
|----|-----|
|pods |	Total number of pods|
|replicationcontrollers |	Total number of replication controllers|
|services |	Total number of services|
|secrets |	Total number of secrets|
|persistentvolumeclaims |	Total number of persistent volume claims|

The following table describes some compute resources that can be restricted by a quota:

|Compute Resource Name|	Quota Description|
|---|----|
|cpu (requests.cpu)	Total |CPU use across all containers|
|memory (requests.memory)	|Total memory use across all containers|
|storage (requests.storage)|	Total storage requests by containers across all persistent volume claims|

Quota attributes can track either resource requests or resource limits for all pods in the project. By default, quota attributes track resource requests. Instead, to track resource limits, prefix the compute resource name with limits, for example, limits.cpu.

The following listing show a ResourceQuota resource defined using YAML syntax. This example specifies quotas for both the number of resources and the use of compute resources:

> ```

Reviewer 2: Syntax highlighting 
```suggestion
```yaml
```

compute-long-running-quota   2d23h   pods: 2/100, requests.cpu: 60m/500m, requests.memory: 276Mi/2Gi   limits.cpu: 330m/1500m, limits.memory: 1124Mi/4Gi
compute-time-bound-quota     2d23h   pods: 0/100, requests.cpu: 0/500m, requests.memory: 0/2Gi         limits.cpu: 0/1500m, limits.memory: 0/4Gi
```

Without arguments, the `oc describe quota` command displays the cumulative limits set for all ResourceQuota resources in the project:

```
Name:       compute-long-running-quota
Namespace:  ad204f-dev
Scopes:     NotBestEffort, NotTerminating
 * Matches all pods that have at least one resource requirement set. These pods have a burstable or guaranteed quality of service.
 * Matches all pods that do not have an active deadline. These pods usually include long running pods whose container command is not expected to terminate.
Resource         Used    Hard
--------         ----    ----
limits.cpu       330m    1500m
limits.memory    1124Mi  4Gi
pods             2       100
requests.cpu     60m     500m
requests.memory  276Mi   2Gi
```
Also getting the yaml output of a quota will show the status of the quota.

> ```

Reviewer 2: Syntax highlighting
```suggestion
```yaml
```

| Scope | Description|
|---|---|
|Terminating | Match pods where spec.activeDeadlineSeconds >= 0.|
| NotTerminating | Match pods where spec.activeDeadlineSeconds is nil.|
| BestEffort | Match pods that have best effort quality of service for either cpu or memory.|
| NotBestEffort| Match pods that do not have best effort quality of service for cpu and memory.|

You'll notice there are a few quotas within the BC Gov OpenShift projects. Have a look at quota details and see how much has been consumed.
## Limit Ranges

A LimitRange resource, also called a limit, defines the default, minimum, and maximum values for compute resource requests, and the limits for a single pod or container defined inside the project. 

To understand the difference between a limit range and a resource quota, consider that a limit range defines valid ranges and default values for a single container, and a resource quota defines only top values for the sum of all pods in a project. 

A cluster administrator concerned about resource usage in an OpenShift cluster usually defines both limits and quotas for a project.

> A limit range resource can also define default, minimum, and maximum values for the storage capacity requested by an image, image stream, or persistent volume claim. 

Reviewer 2: How do you set defaults for storage capacity for Images?

Author: Whoop. Typo has been fixed now.
